src/compose/MentionWarnings.js

```diff
@@ -0,0 +1,140 @@
+/* @flow strict-local */
+
+import React, { PureComponent } from 'react';
+import { connect } from 'react-redux';
+
+import type { Auth, Stream, Dispatch, Narrow, UserOrBot, Subscription } from '../types';
+import { getActiveUsersById, getAuth } from '../selectors';
+import { isPrivateNarrow } from '../utils/narrow';
+import * as api from '../api';
+
+import AnimatedScaleComponent from '../animation/AnimatedScaleComponent';
+import MentionedUserNotSubscribed from '../message/MentionedUserNotSubscribed';
+
+type State = {|
+  unsubscribedMentions: UserOrBot[],
+|};
+
+type SelectorProps = {|
+  auth: Auth,
+  usersById: Map<number, UserOrBot>,
+|};
+
+type Props = $ReadOnly<{|
+  narrow: Narrow,
+  stream: Subscription | {| ...Stream, in_home_view: boolean |},
+
+  dispatch: Dispatch,
+  ...SelectorProps,
+|}>;
+
+class MentionWarnings extends PureComponent<Props, State> {
+  state = {
+    unsubscribedMentions: [],
+  };
+
+  /**
+   * Check whether the message text entered by the user contains
+   * an @-mention to a user unsubscribed to the current stream, and if
+   * so, shows a warning.
+   *
+   * This function is expected to be called by `ComposeBox` using a ref
+   * to this component.
+   *
+   * @param message The message text entered by the user in the compose box.
+   */
+  handleMentionSubscribedCheck = async (message: string) => {
+    const { usersById, narrow, auth, stream } = this.props;
+
+    if (isPrivateNarrow(narrow)) {
+      return;
+    }
+    const unformattedMessage = message.split('**')[1];
```

Comment:
This logic looks quite suspicious to me.

It seems like we're trying to parse the message's markup to look for an @-mention. That's a tricky thing to do accurately.

If we were to write logic to do that, it should be in a function of its own with a nice clean interface, and with thorough unit tests. For example: what about a message that has some `**bold**` text and then has an @-mention? Or, what about a message that has several @-mentions in it?

Looking at the place this is called, though, it looks like the argument it's passed may not actually be the message text? That's contrary to this function's jsdoc if so. But `handleMessageAutocomplete` passes it the `completion` it received, which in turn is described at the `onAutocomplete` prop's jsdoc as:
```
   * @param completion The suggestion selected by the user.
```
Which, I'm still not 100% sure what that means -- does it look like `@**Divyanshu Agrawal**` or `Divyanshu Agrawal` or something else? (Based on the code here in this function, I'm guessing the former.) Either way, that makes this problem a lot more tractable than it would be if it were really an entire message.

Reply:
Oh yeah - the jsdoc is incorrect. Here, i'm only removing the `**` from the @-mention, not the complete message.

Reply:
I've added relevant JSDocs to make it more clear what the `completion` looks like. ( It's of the form `**Divyanshu Agrawal**` ).

Reply:
> It's of the form `**Divyanshu Agrawal**`

Neither of the things I guessed! Glad the answer is written down :wink: